src/osd/scrubber/scrub_reservations.cc

```diff
@@ -194,20 +230,30 @@ void ReplicaReservations::verify_rejections_source(
 		  *op->get_req())
 	   << dendl;
 
+  // is this a stale response to a previous request (e.g. one that
+  // timed-out)? if so - ignore it (but do log an error, to cause tests
+  // to fail)
+  const auto msg_epoch = op->get_req<MOSDScrubReserve>()->map_epoch;
+  const auto msg_nonce = op->get_req<MOSDScrubReserve>()->reservation_nonce;
+  const auto nonce_verified = is_response_relevant(msg_epoch, msg_nonce, from);
+  if (!nonce_verified) {
+    m_osds->clog->warn() << nonce_verified.error();
+    return false;
+  }
+
+  log_failure_and_duration(scrbcnt_resrv_rejected);
+  ceph_assert(get_last_sent().has_value());
   // verify that the denial is from the peer we expected. If not?
   // we should treat it as though the *correct* peer has rejected the request,
   // but remember to release that peer, too.
-
-  ceph_assert(get_last_sent().has_value());
-  const auto expected = *get_last_sent();
-  if (from != expected) {
-    dout(1) << fmt::format(
-		   "unexpected rejection from {} (expected {})", from, expected)
-	    << dendl;
+  const auto peer_verified = is_msg_source_correct(msg_epoch, from);
+  if (!peer_verified) {
+    m_osds->clog->warn() << peer_verified.error();
```

Comment:
This should actually be impossible, right?  If so, please explain that in the comment.

Reply:
added a comment line

Reply:
Where?  Having trouble finding it.

Reply:
Sorry - I wasn't clear. I am making changes and acknowledging your comments one by one. Will push a new version today.

Reply:
No worries, I'll check back in tomorrow.

Reply:
@athanatos - I think I got all fixes right now.